### PR TITLE
[performance] ensure `causal_mask` is created directly on device

### PR DESCRIPTION
**Pull request by jeffra**
# What does this PR do?

@tjruwase and @tohtana discovered that causal_mask is currently being created on CPU then moved to GPU during the forward pass of OPT (and we think other models). This appears to be causing a significant performance degradation on multi-gpu environments due to parallel host to device copies going on. It's not 100% clear to us why this is so bad but here is what we observe before and after this patch:

Before this patch w. OPT-125m on x8 A100s:
image

After the patch:
image

These numbers were gathered from a modified version of https://github.com/huggingface/transformers/blob/main/examples/pytorch/language-modeling/run_clm.py but turning on `wall_clock_breakdown: true` in our deepspeed config.

One major complication we see in accepting this PR is that the two functions being modified are copied across lots of different models and the `make fix-copies` script doesn't seem to address all of them correctly across both `_make_causal_mask` and `_prepare_decoder_attention_mask`

## Who can review?

Tagging @sgugger and @stas00 to help triage to the right people

### CODE REVIEW

**Comment by ydshieh** [comment]
cc @thomasw21 @NouamaneTazi since both of you are experts on this kind of things - to see if you have any general opinion and/or if you would like to review this PR too.

**Comment by ydshieh** [comment]
@jeffra Would it possible for you (and/or @tjruwase and @tohtana) to provide your script that finds/measures/profiles the running time for this issue 🙏 . It would be super helpful for us to dive into internally too.

File: src/transformers/models/opt/modeling_opt.py [hunk 1]
```diff
@@ -64,18 +64,20 @@
...
+    input_ids_shape: torch.Size, dtype: torch.dtype, device: torch.device, past_key_values_length: int = 0
+):
     """
     Make causal mask used for bi-directional self-attention.
     """
     bsz, tgt_len = input_ids_shape
-    mask = torch.full((tgt_len, tgt_len), torch.tensor(torch.finfo(dtype).min))
-    mask_cond = torch.arange(mask.size(-1))
+    mask = torch.full((tgt_len, tgt_len), torch.tensor(torch.finfo(dtype).min, device=device), device=device)
+    mask_cond = torch.arange(mask.size(-1), device=device)
```
Comment (NouamaneTazi) [inline src/transformers/models/opt/modeling_opt.py:75]:
```suggestion
    mask = torch.full((tgt_len, tgt_len), torch.finfo(dtype).min, device=device)
    mask_cond = torch.arange(mask.size(-1), device=device)
```
It seems that `torch.tensor(torch.finfo(torch.float32).min, device="cuda")` requires CPU-GPU sync

Comment (NouamaneTazi) [reply src/transformers/models/opt/modeling_opt.py:75]: Otherwise if a tensor is needed, we can also do
```
torch.ones(1, device="cuda") * torch.finfo(torch.float32).min # no sync
```

**Review by nmntz — COMMENTED** [review]
I'm a big supporter of removing CPU-GPU syncs, so I would very like see this merged! ⚡️

**Review by sgugger — APPROVED** [review]
LGTM, thanks a lot for the fix! Note that the same modification needs to be applied to BART (since OPT copies from BART) in order for all quality checks to pass.

**Comment by sgugger** [comment]
I think just copying the same fix to BART and then applying `make fix-copies` is simple enough for this PR. Dealing with functions that are not copies or are named differently can indeed be done in followup PRs.

**Comment by jeffra** [comment]
Ok, i've updated the BART implementation and attempted again to get `make fix-copies` to work for me but I think I might be doing something wrong. Some of the original issues I saw are now fixed on other models (e.g., https://github.com/huggingface/transformers/pull/22382 adds a `# Copied from` tag for llama). However, I am still seeing issues i think coming from the fix-up scripts getting confused with the function signature change of `_make_causal_mask`.

Can someone try `make fix-copies` on their side with this? You should be able to push to my branch.

**Comment by sgugger** [comment]
Ah yes, `make fix-copies` does not change the signature of the function so that is indeed something to edit manually. If it's too much work I can try to push this to your branch tomorrow.

**Comment by jeffra** [comment]
> Ah yes, `make fix-copies` does not change the signature of the function so that is indeed something to edit manually. If it's too much work I can try to push this to your branch tomorrow.

Sounds good, I might have some time this afternoon for this. Otherwise feel free to do it :)

**Comment by jeffra** [comment]
Okay all the models should be fixed now, `make fixup` is clear on my local tests.

File: src/transformers/models/opt/modeling_opt.py (hunk 1 above)
Comment (NouamaneTazi) [reply src/transformers/models/opt/modeling_opt.py:76]: I just tried my suggestion:`mask = torch.full((tgt_len, tgt_len), torch.finfo(dtype).min, device=device)` 
and it does work on "cpu". I'm not sure why it wouldn't work? 🤔

Comment (thomasw21) [reply src/transformers/models/opt/modeling_opt.py:76]: From the pytorch doc (https://pytorch.org/docs/stable/generated/torch.full.html), `torch.full` allows for scalar type for `fill_value`. So I think what @NouamaneTazi is trying to convey is that you don't need to put it first cast `-inf` to a tensor type and then fill, you could just fill with `-inf`

**Review by sgugger — APPROVED** [review]
Thanks for copy-pasting the signature changes manually!
